apps/platform/src/lists/ListService.ts

```diff
@@ -170,21 +171,45 @@ export const populateList = async (list: List, rule: Rule) => {
     const version = oldVersion + 1
     await updateList(id, { state: 'loading', version })
 
-    type UserListChunk = { user_id: number, list_id: number, version: number }
-    const query = ruleQuery(rule, list.project_id)
-    await chunk<UserListChunk>(query, 100, async (items) => {
-        if (items.length <= 0) return
+    // look up all users in list's project
+    const query = User.query().where('project_id', list.project_id)
+
+    // collect matching user ids, insert in batches of 100
+    const chunker = new Chunker<number>(async userIds => {
         await UserList.query()
-            .insert(items)
+            .insert(userIds.map(user_id => ({
+                list_id: list.id,
+                user_id,
+                version,
+            })))
             .onConflict(['user_id', 'list_id'])
             .merge(['version'])
-    }, ({ id: user_id }: { id: number }) => ({ user_id, list_id: id, version }))
+    }, 100)
+
+    // stream all users, check rules in batches
+    await chunk<User>(query, 100, async users => {
+        const events = await getUserEventsForRules(users.map(u => u.id), [rule])
+            .then(events => groupBy(events, e => e.user_id))
+        for (const user of users) {
+            const matched = check({
+                user: user.flatten(),
+                events: events.get(user.id) ?? [],
+            }, rule)
+            if (matched) {
+                chunker.add(user.id)
+            }
+        }
+    })
```

Comment:
This is quite a piece of insanity. Would probably vote to have it send this off to jobs in chunks instead of having what could be a tens of millions of users request (and their subsequent events) taking up the entire event queue. Not to mention getting a users events if somewhat blocking since it's going to wait until it has every event to proceed instead of streaming them